src/controller/buffer-controller.ts

```diff
@@ -1,55 +1,102 @@
+  // cache the self generated object url to detect hijack of video tag
+  private _objectUrl: string | null = null;
+
+  // signals that the sourceBuffers need to be flushed
+  private _needsFlush: boolean = false;
+
+  // signals that mediaSource should have endOfStream called
+  private _needsEos: boolean = false;
+
+  // this is optional because this property is removed from the class sometimes
+  public audioTimestampOffset?: number;
+
+  // The number of BUFFER_CODEC events received before any sourceBuffers are created
+  public bufferCodecEventsExpected: number = 0;
+
+  // A reference to the attached media element
+  public media: HTMLMediaElement | null = null;
+
+  // A reference to the active media source
+  public mediaSource: MediaSource | null = null;
+
+  // List of pending segments to be appended to source buffer
+  public segments: Segment[] = [];
+
+  public parent?: string;
+
+  // A guard to see if we are currently appending to the source buffer
+  public appending: boolean = false;
+
+  // counters
+  public appended: number = 0;
+  public appendError: number = 0;
+  public flushBufferCounter: number = 0;
+
+  public tracks: TrackSet = {};
+  public pendingTracks: TrackSet = {};
+  public sourceBuffer: SourceBufferSet = { };
+  public flushRange: SourceBufferFlushRange[] = [];
+
+  constructor (hls: any) {
     super(hls,
-      Event.MEDIA_ATTACHING,
-      Event.MEDIA_DETACHING,
-      Event.MANIFEST_PARSED,
-      Event.BUFFER_RESET,
-      Event.BUFFER_APPENDING,
-      Event.BUFFER_CODECS,
-      Event.BUFFER_EOS,
-      Event.BUFFER_FLUSHING,
-      Event.LEVEL_PTS_UPDATED,
-      Event.LEVEL_UPDATED);
-
-    // the value that we have set mediasource.duration to
-    // (the actual duration may be tweaked slighly by the browser)
-    this._msDuration = null;
-    // the value that we want to set mediaSource.duration to
-    this._levelDuration = null;
-    // the target duration of the current media playlist
-    this._levelTargetDuration = 10;
-    // current stream state: true - for live broadcast, false - for VoD content
-    this._live = null;
-    // cache the self generated object url to detect hijack of video tag
-    this._objectUrl = null;
-    // The number of BUFFER_CODEC events received before any sourceBuffers are created
-    this.bufferCodecEventsExpected = 0;
-
-    // Source Buffer listeners
-    this.onsbue = this.onSBUpdateEnd.bind(this);
-    this.onsbe = this.onSBUpdateError.bind(this);
-    this.pendingTracks = {};
-    this.tracks = {};
+      Events.MEDIA_ATTACHING,
+      Events.MEDIA_DETACHING,
+      Events.MANIFEST_PARSED,
+      Events.BUFFER_RESET,
+      Events.BUFFER_APPENDING,
+      Events.BUFFER_CODECS,
+      Events.BUFFER_EOS,
+      Events.BUFFER_FLUSHING,
+      Events.LEVEL_PTS_UPDATED,
+      Events.LEVEL_UPDATED);
   }
 
   destroy () {
     EventHandler.prototype.destroy.call(this);
   }
 
-  onLevelPtsUpdated (data) {
+  onLevelPtsUpdated (data: any) {
```

Comment:
looks simple to add a type for this
```suggestion
  onLevelPtsUpdated (data: { start: number }) {
```